c/examples/cpp_sorting_example.cpp

```diff
@@ -0,0 +1,121 @@
+    //        "the sorter does not currently handle edge metadata");
+    //}
+    if (start != 0) {
+        throw std::invalid_argument("the sorter requires start==0");
+    }
+
+    std::vector<edge_plus_time> temp;
+    temp.reserve(static_cast<std::size_t>(sorter->tables->edges.num_rows));
+
+    const auto *edges = &sorter->tables->edges;
+    const auto *nodes = &sorter->tables->nodes;
+
+    // NOTE: this future-proofs the type against changing to int64_t
+    // for tsk_id_t.  It will also look odd to a C programmer,
+    // but the edges->parent[0] need not be valid at run time.
+    // Rather, it is checking what the return type of the expression
+    // would be, which is tsk_id_t& (with the & being due to decltype
+    // not returning the decay type).
+    const auto num_nodes
+        = static_cast<std::remove_reference<decltype(edges->parent[0])>::type>(
+            nodes->num_rows);
+
+    for (decltype(sorter->tables->edges.num_rows) i = 0;
+         i < sorter->tables->edges.num_rows; ++i) {
+        if (edges->parent[i] >= num_nodes || edges->child[i] >= num_nodes) {
+            throw std::invalid_argument("edge table contains invalid node ids");
+        }
+        temp.push_back(edge_plus_time{ nodes->time[edges->parent[i]], edges->parent[i],
+            edges->child[i], edges->left[i], edges->right[i] });
+    }
+
+    if (temp.size() != static_cast<std::size_t>(edges->num_rows)) {
+        throw std::runtime_error("fatal error: incorect number of copied edges!");
+    }
+
+    std::sort(begin(temp), end(temp),
+        [](const edge_plus_time &lhs, const edge_plus_time &rhs) {
+            if (lhs.time == rhs.time) {
+                if (lhs.parent == rhs.parent) {
+                    if (lhs.child == rhs.child) {
+                        return lhs.left < rhs.left;
+                    }
+                    return lhs.child < rhs.child;
+                }
+                return lhs.parent < rhs.parent;
+            }
+            return lhs.time < rhs.time;
+        });
+
+    // NOTE: at this point, exceptions are no longer
+    // possible.
+
+    for (std::size_t i = 0; i < temp.size(); ++i) {
+        edges->left[i] = temp[i].left;
+        edges->right[i] = temp[i].right;
+        edges->parent[i] = temp[i].parent;
+        edges->child[i] = temp[i].child;
+    }
+
+    return 0;
+}
+
+int
+main(int argc, char **argv)
+{
+    if (argc != 3) {
+        std::cerr << "Usage: " << argv[0] << " input.trees output.trees\n";
+        std::exit(0);
+    }
+    const char *infile = argv[1];
+    const char *outfile = argv[2];
+
+    tsk_table_collection_t tables;
+    auto ret = tsk_table_collection_load(&tables, infile, 0);
+    handle_tskit_return_code(ret);
+
+    tsk_table_sorter_t sorter;
+    ret = tsk_table_sorter_init(&sorter, &tables, 0);
+    handle_tskit_return_code(ret);
+    sorter.sort_edges = sort_edges;
+    try {
+        ret = tsk_table_sorter_run(&sorter, NULL);
```

Comment:
Cool, so exceptions will propagate up through the stack? That's very handy.

Reply:
Yes, they'll propagate.  C++ is responsible for the stack here.